source/agora/test/QuorumPreimage.d

```diff
@@ -0,0 +1,424 @@
+import agora.consensus.data.Block;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import agora.consensus.Genesis;
+import agora.node.FullNode;
+import agora.test.Base;
+
+import std.algorithm;
+import std.format;
+import std.range;
+
+import core.thread;
+import core.time;
+
+/// test preimage changing quorum configs
+unittest
+{
+    import agora.common.Types;
+    TestConf conf = { nodes : 6, max_listeners : 7,
+        topology : NetworkTopology.TwoOutsiderValidators,
+        extra_blocks : 8,
+        validator_cycle : 10 };
+    auto network = makeTestNetwork(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 8, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 8)));
+
+    const quorums_1 = [
+        // 0
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 1
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 2
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 3
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 4
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 5
+        QuorumConfig(5, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        QuorumConfig.init,
+        QuorumConfig.init,
+    ];
+
+    version (none)
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getQuorumConfig() == quorums_1[idx], 5.seconds,
+            format("Node %s has quorum config %s. Expected: %s",
+                idx, node.getQuorumConfig(), quorums_1[idx])));
+
+    Transaction makePayTx (in Transaction prev, PublicKey[] addresses, uint index = 0)
```

Comment:
I guess this should use the new TxBuilder routines. I haven't used them yet personally.

Reply:
They are still there tho

Reply:
I'll remove these.